### ISSUE TITLE
EventReceiver mangles timestamps

### ISSUE THREAD

**Issue by mgedmin** [celery/celery#2482]
1. Python's [time.time()](https://docs.python.org/3/library/time.html#time.time) returns the number of seconds since the [epoch](https://docs.python.org/3/library/time.html#module-time) (Jan 1, 1970, midnight UTC), often called a time_t value.
2. celery.events.EventDispatcher.publish() creates an event dictionary with a `timestamp` item containing [the time_t of the event creation](https://github.com/celery/celery/blob/1380902bb4e75d9c5aed4954032442339620de05/celery/events/__init__.py#L59), and a `utcoffset` that contains the [UTC offset](https://github.com/celery/celery/blob/1380902bb4e75d9c5aed4954032442339620de05/celery/events/__init__.py#L193) of the local timezone, in hours ([truncated](https://github.com/celery/celery/blob/1380902bb4e75d9c5aed4954032442339620de05/celery/utils/timeutils.py#L358-L361), which mishandles time zones like India's +05:30).
3. celery.events.EventReceiver.event_from_message() takes the `timestamp` and `utcoffset` values from an event dictionary and [calls](https://github.com/celery/celery/blob/1380902bb4e75d9c5aed4954032442339620de05/celery/events/__init__.py#L386-L392) `adjust_timestamp(timestamp, utcoffset)`, which *deep breath* takes the difference between the local system's UTC offset and the event's UTC offset and [subtracts](https://github.com/celery/celery/blob/1380902bb4e75d9c5aed4954032442339620de05/celery/utils/timeutils.py#L364-L365) that difference (converted to seconds) from the time_t value, resulting in something that doesn't quite correspond to any physical quantity.
4. celery.events.EventReceiver.event_from_message() also calls `time.time()` and [puts](https://github.com/celery/celery/blob/1380902bb4e75d9c5aed4954032442339620de05/celery/events/__init__.py#L393) the time_t value into `local_received`
5. celery.events.state.Worker.event takes an (absolute) difference between `local_received` and `timestamp` (adjusted in step 3) and [complains](https://github.com/celery/celery/blob/1380902bb4e75d9c5aed4954032442339620de05/celery/events/state.py#L165-L168) about clock drift if the difference is greater than 16 seconds (HEARTBEAT_DRIFT_MAX).  But the difference is now `timestamp1_in_utc - timestamp2_in_utc + strange_utc_offset_adjustment`, instead of a straightforward `timestamp1_in_utc - timestamp2_in_utc`!

I humbly submit that something's not right here.  I don't think EventReceiver should ever tamper with time_t values.

**Comment by sorreltree** [comment]
Just want to add my +1 that EventReceiver shouldn't be munging timestamp values.  If you're going to use the time.time() value, it really should never be touched.

Times in python are a royal pain -- the only sane practice I've come to use is what's recommended by the pytz documentation -- move everything to explicit (and correct!) UTC as soon as you get a value, and hold it as UTC until the moment you're ready to write it out (as a string or a timestamp or whatever).  Since AMQP wants a timestamp as seconds since the epoch (which is defined in UTC), carrying that timestamp around should also be fine, but again, it should be localized using something like pytz.timezone(tz_string).normalize(), not by adding a timezone offset to it.

**Comment by ask** [comment]
We are not expecting the timestamp to be useful for anything since this is a networked system, so if it's off by seconds, or minutes it shouldn't matter.

The logical clock value is used as primary key for ordering of events, then the hostname, and last
the timestamp.  So the timestamp is really only used for old event producers, and as a key for clients without a clock value (like task producers usually).

ETA/countdown tasks do depend on wall clock time so I added the drift warning there to have a notification if the clocks could be out of sync.

The offset was a quick hack to keep backwards compatibility, I tried using mktime(datetime.utctimetuple()) but that had a big impact on performance and
using mktime means rounding up to the nearest second.

We could possibly be using ISO8601 strings instead, as the rest of Celery uses, but I'm not
sure how that works as a key in the ordering of events for clock-less event producers.

I guess we could also write a custom mktime to take microseconds into account.
